net: lwm2m: Allow Bootstrap server to close DTLS connection

SeppoTakalo opened the pull request:
Allow Bootstrap server to close the DTLS connection immediately after receiving Ack to Bootstrap-Finnish command.
This is not an error as either parties are allowed to tear down the connection.

This behaviour was seen they tried Zephyr's client against IoTerop’s ALASKA server.

![Screenshot from 2023-09-12 11-37-37](https://github.com/zephyrproject-rtos/zephyr/assets/3104794/2e939d6a-f262-428e-8941-2704c75d4749)

SeppoTakalo committed:
net: lwm2m: Allow Bootstrap server to close DTLS connection

Allow Bootstrap server to close the DTLS connection immediately
after receiving Ack to Bootstrap-Finish command.
This is not an error as either parties are allowed to tear down
the connection.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

jukkar commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -351,14 +351,20 @@ static void sm_handle_failure_state(enum sm_engine_state sm_state)
 static void socket_fault_cb(int error)
```
Typo s/Finnish/Finish/ in commit msg

jukkar commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -351,14 +351,20 @@ static void sm_handle_failure_state(enum sm_engine_state sm_state)
-	lwm2m_socket_close(client.ctx);
+		if (get_sm_state() == ENGINE_BOOTSTRAP_TRANS_DONE) {
+			/* Ignore the error, some servers close the connection immediately
+			 * after receiving Ack to Bootstra-Finnish command.
```
and here

SeppoTakalo commented:
Fixed the typo.

I also removed the `#ifdef ...BOOSTRAP` around the enum that defines bootstrap states. It is much cleaner to write the checks using `IS_ENABLED()` instead of if-deffing various cases, because those enum values don't exist.

jukkar reviewed (approved):
LGTM
